src/main/scala/za/co/absa/hyperdrive/trigger/scheduler/eventProcessor/EventProcessor.scala

```diff
@@ -42,27 +42,66 @@ class EventProcessor(
 
   private def processEvents(events: Seq[Event], sensorId: Long, triggeredBy: String)(
     implicit ec: ExecutionContext
-  ): Future[Boolean] =
+  ): Future[Boolean] = {
+    logger.trace(
+      "Processing events {} called on event processor for sensor (SensorId={}), triggered by: {}",
+      new LazyToStr(events.map(e => s"EventId=${e.id}")),
+      sensorId,
+      triggeredBy
+    )
     eventRepository.getExistEvents(events.map(_.sensorEventId)).flatMap { eventsIdsInDB =>
       val newEvents = events.filter(e => !eventsIdsInDB.contains(e.sensorEventId))
+      logger.trace(
+        "Unprocessed events {}",
+        new LazyToStr(newEvents.map(e => s"EventId=${e.id}"))
+      )
       if (newEvents.nonEmpty) {
         dagDefinitionRepository.getJoinedDagDefinition(sensorId).flatMap {
           case Some(joinedDagDefinition) =>
             for {
-              hasInQueueDagInstance <- dagInstanceRepository.hasInQueueDagInstance(joinedDagDefinition.workflowId)
+              hasInQueueDagInstance <- dagInstanceRepository
```

Comment:
You can rewrite this and have logging after each "flatmap"
_ = logger.trace("bla")

Reply:
Yeah, I know about it, and I will change it back. However, I don't like it because the `=` is for a map if scala had something for specifying side effects in for comprehension.

I would instead use something like [https://typelevel.org/cats/typeclasses/arrow.html](https://typelevel.org/cats/typeclasses/arrow.html) or a better example with precisely the same thing [https://medium.com/virtuslab/arrows-monads-and-kleisli-part-ii-12ffd4da8bc9](https://medium.com/virtuslab/arrows-monads-and-kleisli-part-ii-12ffd4da8bc9).

To me, writing it that way looks much cleaner, and also, using cats would add several additional structures, like  `IO`, `Functor` (which in turn can be used to simplify all the `.map(_.map(..._.map(f)...))` operations), `Validated`, `Writer`, etc...